### PR TITLE
refactoring RoadmapCanvas

### PR DESCRIPTION
**Pull request by atago0129**
RoadmapCanvas 内の処理がべた書きだったのをせめてメソッドに分割して多少可読性を上げた（つもり）

### CODE REVIEW

File: src/Roadmappy.js [hunk 1]
```diff
@@ -9,8 +9,7 @@ export class Roadmappy {
 
   constructor(options, {tasks, people}) {
     this.roadmapCanvas = new RoadmapCanvas(options);
-    let parser = new RoadmapParser();
-    this.roadmap = parser.parse(tasks, people);
+    this.roadmap = new RoadmapParser().parse(tasks, people);
```
Comment (hrsh7th) [inline src/Roadmappy.js]: nice.

File: src/Roadmappy.js [hunk 2]
```diff
@@ -9,8 +9,7 @@ export class Roadmappy {
 
   constructor(options, {tasks, people}) {
     this.roadmapCanvas = new RoadmapCanvas(options);
```
Comment (hrsh7th) [inline src/Roadmappy.js]: It's name bit ugly.

Should change `roadmapCanvas` to `canvas`.

File: src/Roadmappy.js (hunk 1 above)
Comment (hrsh7th) [reply src/Roadmappy.js]: But should change args count 2 to 1.

File: src/component/Roadmap.js [hunk 3]
```diff
@@ -11,11 +11,65 @@ export class Roadmap {
...
 
   getTasks() {
-    return this.tasks;
+    return this._sortItems(this.tasks);
   }
 
   getPeople() {
-    return this.people;
+    return this._sortItems(this.people);
+  }
```
Comment (hrsh7th) [inline src/component/Roadmap.js]: `sortItems` accepts tasks or peoples, but it's implicitly.

Should create `AbstractItem` class.
Tasks extends AbstractItem.
Person extends AbstractItem.

What do you think?

File: src/component/Roadmap.js [hunk 4]
```diff
@@ -11,11 +11,65 @@ export class Roadmap {
...
+        }
+        groups.push({
+          type: "group",
+          name: items[i].group,
+          count: count,
+          style: items[i].style
+        });
+      }
+    }
+    return groups;
```
Comment (hrsh7th) [inline src/component/Roadmap.js]: should simplify.

File: src/component/RoadmapCanvas.js [hunk 5]
```diff
@@ -153,7 +131,9 @@ export class RoadmapCanvas {
         })
       ])
       .rangeRound([0, w - sidePadding - 15]);
+  }
 
+  _drwanXAxis(svg, xScale, sidePadding, topPadding) {
```
Comment (hrsh7th) [inline src/component/RoadmapCanvas.js]: drawn?

File: src/component/RoadmapCanvas.js [hunk 6]
```diff
@@ -3,101 +3,69 @@ import ContextMenu from "d3-v4-contextmenu";
...
+
+    let xScale = this._generateXScale(items, w, sidePadding);
+    this._drwanXAxis(svg, xScale, sidePadding, this.style.topPadding);
+
+    this._drawItemLines(items, svg, xScale, this.style.gap, sidePadding, this.style.topPadding, this.style.barHeight);
+
+    this._addMouseHelper(svg, xScale, this.style.barHeight, sidePadding, roadmap);
+  }
+
+  _generateSVG(w, h) {
```
Comment (hrsh7th) [inline src/component/RoadmapCanvas.js]: Should do only `generate`, Current implmentation has 2 feature `generate` and `append`.

File: src/component/RoadmapCanvas.js [hunk 7]
```diff
@@ -3,101 +3,69 @@ import ContextMenu from "d3-v4-contextmenu";
...
-    const h = items.length * gap + 40;
+  _render(items, groups, roadmap) {
     const w = this.targetElement.node().clientWidth;
+    const h = items.length * this.style.gap + 40;
 
-    let svg = this.targetElement.append("svg").attr("width", w).attr("style", "overflow: visible");
-    svg.attr("height", function() {
-      return parseInt(svg.attr("height") || 0, 10) + h;
-    });
+    let svg = this._generateSVG(w, h);
```
Comment (hrsh7th) [inline src/component/RoadmapCanvas.js]:
```
let svg = this._generageSVG(w, h);
...
this._appendHoge(svg);
...
this.targetElement.appendChild(svg);
```

**Comment by atago0129** [comment]
I decided to create class of task, assignee and story.
And I did clean up code.

File: src/component/RoadmapCanvas.js (hunk 7 above)
Comment (atago0129) [reply src/component/RoadmapCanvas.js]: i will try

File: src/component/RoadmapCanvas.js [hunk 8]
```diff
@@ -3,101 +3,134 @@ import ContextMenu from "d3-v4-contextmenu";
...
 
-  _render(items, roadmap) {
-    const barHeight = 20;
-    const gap = barHeight + 4;
-    const topPadding = 20;
+  /**
+   * @param {Roadmap} roadmap
+   * @private
+   */
+  _renderStoryTask(roadmap) {
```
Comment (atago0129) [inline src/component/RoadmapCanvas.js]: `_renderStoryTask` and `_renderAssigneeTask` have duplicate codes...
But, I can't think of good solution.

File: src/component/RoadmapParser.js [hunk 9]
```diff
@@ -1,57 +1,113 @@
...
+   */
+  _parseStories(stories) {
+    for (let i = 0; i < stories.length; i++) {
+      let story = stories[i];
+      if (story.id === undefined || story.name === undefined) {
+        // invalid item
+        continue;
       }
-      roadmap.setTask(task);
+      this.roadmap.setStory(new RoadmapStory(
```
Comment (hrsh7th) [inline src/component/RoadmapParser.js]: should change this.

1. `setStory` -> `addStory`
or
2. `setStory(storyId, new Story(...));`

My recommendation is 1.

Comment (atago0129) [reply src/component/RoadmapParser.js]: yes. i will change 1

File: src/component/RoadmapTask.js [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+   * @param {number} order
+   * @param {string} from
+   * @param {string} to
+   * @param {number} involvement
+   */
+  constructor(id, name, storyId, assigneeIdList, color, order, from, to, involvement) {
+    this.id = id;
+    this.name = name;
+    this.storyId = storyId;
+    this.assigneeIdList = assigneeIdList;
```
Comment (hrsh7th) [inline src/component/RoadmapTask.js]: It should convert to Array.
Use this idiom.

```
this.assigneeIdList = [].concat(arrayOrStringOrUndefined || []);
```

Comment (atago0129) [reply src/component/RoadmapTask.js]: thanks!

File: src/component/RoadmapCanvas.js (hunk 7 above)
Comment (atago0129) [reply src/component/RoadmapCanvas.js]: i tried to change, but it's too hard...
because i think `d3.js` is not suitable for your suggestion

**Comment by hrsh7th** [comment]
very nice refactor!
I think, We should merge currently this.

We can refactor later.

File: src/component/RoadmapCanvas.js (hunk 7 above)
Comment (hrsh7th) [reply src/component/RoadmapCanvas.js]: sorry... 😭

**Comment by atago0129** [comment]
thanks!!
ma-shima!

**Comment by hrsh7th** [comment]
haha.

I will join coding.

**Pull request merged**